Add Pull of Base Level + Support for Multiple Towers

GCRA101 opened the pull request:
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Fixes #507 

<img width="1515" height="704" alt="Screenshot 2025-10-28 200623" src="https://github.com/user-attachments/assets/18828bf0-dae7-4078-a611-32347fcffc30" />



ETABS Toolkit now can allow to pull all levels (including the Base Level) for all the towers that are defined in the ETABS Model (using the AllowMultipleTower feature).


 ### Test files

Grasshopper File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23512-AddPullBaseLevel+SupportforMultipleTowers/Test%20Script.gh?csf=1&web=1&e=G8rRvd
ETABS File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23512-AddPullBaseLevel+SupportforMultipleTowers/[TestModel_22_MultiTower.EDB](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23512-AddPullBaseLevel+SupportforMultipleTowers/TestModel_22_MultiTower.EDB?csf=1&web=1&e=ydAZhq)?csf=1&web=1&e=ydAZhq


 ### Changelog
- Added Tower fragment
- Added logic in Read/Level.cs to extract info about Base Level + deal with MultiTowers.

GCRA101 committed:
Add Tower Element in ETABS_oM

Tower.cs implements the IFragment interface and allows to store the name of the tower the level belongs to when it gets pulled from the ETABS model.

GCRA101 committed:
Update Level.cs to work with Multiple Towers and pick up the Base Level

The new code gets the list of Towers in the model, extracts the data regarding the base levels of the towers from the Database Tables of the models and return to the user all the levels beloning to all towers.

peterjamesnugent commented on ETABS_oM/Elements/Tower.cs:
```diff
@@ -0,0 +1,51 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace BH.oM.Adapters.ETABS.Elements
```
```suggestion
namespace BH.oM.Adapters.ETABS.Fragments
```

peterjamesnugent commented on Etabs_Adapter/CRUD/Read/Level.cs:
```diff
@@ -47,28 +48,79 @@ public partial class ETABSAdapter : BHoMAdapter
+            string tableKey = "Tower and Base Story Definitions";
+            string[] fieldKeyList = null;
+            string groupName = "All";
+            int tableVersion = 0;
```
These are not used after the `ref`, are they needed?

GCRA101 committed:
Add Alternative Algorithm for ETABS16/17 in Read\Level.cs

The API of ETABS v16 and v17 doesn't have the interfaces cDatabaseTables and cTower that are key to get information on levels (including the base level) when multitower feature is active.

To prevent the code from crashing, whenever the ETABS version run by the user is recognized to be v16 or v17, the original algorithm is used.

This won't return base levels and won't deal with multitowers.

GCRA101 commented:
@BHoMBot check installer

GCRA101 committed:
Remove unnecessary line in Level.cs

GCRA101 commented:
@BHoMBot check installer

samuelbasimalla commented:
Tested this on ETABS Version 22.7.0, Rhino 7

The script works as intended:
1. The base level gets pulled when it is non-zero value (this wasn't the case before, it always pulled base level at "0" m)
2. The levels of multiple towers gets pulled (new feature)


<img width="612" height="722" alt="image" src="https://github.com/user-attachments/assets/b6da576f-d1a0-42d0-88b0-50b8b2ee69d5" />

peterjamesnugent reviewed (changes requested):
Change of namespace - sorry had this on pending

GCRA101 committed:
Update ETABS_oM/Elements/Tower.cs

Change namespace from Elements to Fragments.

Co-authored-by: Peter Nugent <Peter.Nugent@burohappold.com>

GCRA101 committed:
Move Tower.cs from Elements/ to Fragments/

GCRA101 commented:
@peterjamesnugent,
thanks for the comments. Moved Tower.cs from Elements/ to Fragments/
Retested the code and all works fine. 
Updates in latest commit: https://github.com/BHoM/ETABS_Toolkit/pull/512/commits/f6cc38d6901038b970165178e833eed28e791581

GCRA101 commented:
@BHoMBot check installer

peterjamesnugent reviewed (approved):
Changes addressed.

peterjamesnugent reviewed (approved):
Rerequested by mistake - still approved.

peterjamesnugent commented:
@BHoMBot ready-to-merge

peterjamesnugent commented:
@BHoMBot check ready-to-merge
